### DISCUSSION TITLE
How to set automatic recording in headless server?

### DISCUSSION THREAD

**Discussion by openaspace** [jamulussoftware/jamulus#2189] [Technical Support & Help]
Hello I need to set automatic multitrack recording on headless linux server writing files on s3 storj.

[Unit]
Description=server
After=network.target
StartLimitIntervalSec=0

[Service]
Type=simple
User=jamulus
Group=nogroup
NoNewPrivileges=true
ProtectSystem=true
ProtectHome=false
Nice=-20
IOSchedulingClass=realtime
IOSchedulingPriority=0

#### See https://jamulus.io/wiki/Command-Line-Options ####
ExecStart=/bin/sh -c 'exec /usr/bin/jamulus-headless -s -n' 
    --numchannels 16 \
    --recording /media/storj-jamulus
        
Restart=on-failure
RestartSec=30
StandardOutput=journal
StandardError=inherit
SyslogIdentifier=jamulus

[Install]
WantedBy=multi-user.target

**Comment by openaspace** [comment]
I have partially solved with the right command but jamulus report as no existent an S3 mount that the system see correctly


**Dec 27 16:28:35 live jamulus[11542]: '/media/storj-jamulus' does not exist but could not be created.**

**Reply by openaspace** [reply]
I was unable to change owner ship for a s3fs mount.. 
the only way to works is to run jamulus as root process.....

**Reply by rdica** [reply]
Interesting. After reading some documentation on s3fs, I wouldnt use it myself.
s3fs has a 64GB file size limit. And from the man page:
```
BUGS
Due  to  S3's  "eventual consistency" limitations, file creation can and will occasionally
       fail. Even after a successful create, subsequent reads can fail for an indeterminate time,
       even  after  one  or  more  successful  reads.  Create  and read enough files and you will
       eventually encounter this failure. This is not a flaw in s3fs and it is  not  something  a
       FUSE  wrapper  like  s3fs can work around. The retries option does not address this issue.
       Your application must either tolerate or compensate for these  failures,  for  example  by
       retrying creates or reads.
```
Be better to create an EBS volume and mount it. That way you could actually use linux filesystem ACLs.

**Reply by openaspace** [reply]
You need to use the local "cache" option for s3fs...and the bandwidth needs to be fast.
I'm using it to store terabytes from months without problems.

**Reply by gilgongo** [reply]
I use [s3ql ](https://github.com/s3ql/s3ql)myself, which I fine works fine up to files of 2TB. But I also usually create new recordings periodically so as not to let things get too big.

**Reply by rdica** [reply]
Have jamulus write the files to local disk as the jamulus user, then cron a job to write to s3 during off-hours. That way you **don't run jamulus as root**, and you can still use s3 for storage.

I don't see any advantage to having jamulus write directly to s3 in your scenario, I should think heavy network traffic writing to s3 while you're playing/recording would contribute to increased latency and introduce unnecessary cpu and memory resource usage on the server. Even if such is not the case, running jamulus as root is just not a good idea.

**Comment by QJKX** [comment]
I'm guessing you didn't mount the share as the jamulus user? Just changing the permissions isn't enough, FUSE needs to be told to allow other users to access the FUSE mount.

https://stackoverflow.com/questions/17544139/allowing-permission-using-s3fs-bucket-directory-for-other-users

rclone is another alternative to s3fs, but will need a similar setting for FUSE.

**Comment by QJKX** [comment]
Another way of doing it would be to create a unit for the s3 mount that also runs as the jamulus user. You can then use the systemd dependency stuff to make sure it's brought up automatically before the jamulus server.

rclone is nice because you can easily change away from s3 later.
